source/uwp/Renderer/lib/XamlBuilder.h

```diff
@@ -147,17 +65,12 @@ namespace AdaptiveNamespace
         bool m_enableXamlImageHandling = false;
         Microsoft::WRL::ComPtr<ABI::AdaptiveNamespace::IAdaptiveCardResourceResolvers> m_resourceResolvers;
 
-        static Microsoft::WRL::ComPtr<ABI::Windows::UI::Xaml::IUIElement>
-        CreateRootCardElement(_In_ ABI::AdaptiveNamespace::IAdaptiveCard* adaptiveCard,
-                              _In_ ABI::AdaptiveNamespace::IAdaptiveRenderContext* renderContext,
-                              _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs,
-                              Microsoft::WRL::ComPtr<XamlBuilder> xamlBuilder,
-                              _Outptr_ ABI::Windows::UI::Xaml::Controls::IPanel** bodyElementContainer);
-
-        static void ApplyBackgroundToRoot(_In_ ABI::Windows::UI::Xaml::Controls::IPanel* rootPanel,
-                                          _In_ ABI::AdaptiveNamespace::IAdaptiveBackgroundImage* backgroundImage,
-                                          _In_ ABI::AdaptiveNamespace::IAdaptiveRenderContext* renderContext,
-                                          _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs);
+        static HRESULT CreateRootCardElement(_In_ ABI::AdaptiveNamespace::IAdaptiveCard* adaptiveCard,
```

Comment:
>CreateRootCardElement [](start = 23, length = 21)

Is the end game that XamlBuilder ceases to exist altogether (assuming we eventually untangle the image renderer issue)? Wondering whether these top level functions (this and BuildPanelChildren) ultimately belong in XamlHelpers as well (whether or not that's current feasible).

Reply:
Or maybe even on AdaptiveCardRendereComponent (as they do seem more substantial than mere "helpers")

---
In reply to: [325849538](https://github.com/microsoft/AdaptiveCards/pull/3457#discussion_r325849538) [](ancestors = 325849538)

Reply:
I like that idea

---
In reply to: [325849965](https://github.com/microsoft/AdaptiveCards/pull/3457#discussion_r325849965) [](ancestors = 325849965,325849538)

Reply:
that's kinda how I've been thinking about it. after this PR, XamlBuilder's sole reason for existing is pretty much for lifetime management reasons. if we find a way to get rid of that last reason (imaging stuff like you said), the remaining functionality should be moved closer to its use (i.e. `XamlBuilder::BuildImage` -> `AdaptiveImageRenderer::Render`, `XamlBuilder::BuildPanelChildren` -> `XamlHelpers::BuildPanelChildren`)

---
In reply to: [325849965](https://github.com/microsoft/AdaptiveCards/pull/3457#discussion_r325849965) [](ancestors = 325849965,325849538)